Library/Homebrew/cmd/list.rb

```diff
@@ -116,8 +116,16 @@ def list
       ls_args << "-r" if args.r?
       ls_args << "-t" if args.t?
 
-      safe_system "ls", *ls_args, HOMEBREW_CELLAR unless args.cask?
-      list_casks(args: args) unless args.formula?
+      ohai "Formulae" if $stdout.tty? && !args.formula?
+      safe_system "ls", *ls_args, HOMEBREW_CELLAR
+
+      unless args.formula?
```

Comment:
```suggestion
      if !args.formula? && Cask::Caskroom.casks.any?
```

Or

```suggestion
      if !args.formula? && Cask::Caskroom.casks(config: Config.from_args(args))
```

TBH I don't really understand the cask arg handling at the moment so I could be misleading you with that.

This way, though, we don't have to show the cask message if none are installed

Reply:
Great suggestion. Not sure how to do the same for formulae though. I could capture the output of `ls` and check that, but then I don't know how to preserve the output formatting while doing that.

Reply:
Actually, `Formula.installed.any?` Should also work, I think

Reply:
Sorry, missed this comment earlier. I think that's a good option. It's kind of slow to run, though, which isn't ideal, but I'm not sure what a better way to do it would be. I guess you could always have two `ls` calls. One could be a `Utils.popen_read` just to see if there's output. Since it's just an `ls` I'd think that wouldn't add too much overhead.

Reply:
Or, better:
```ruby
Dir[HOMEBREW_CELLAR].empty?
```

Reply:
👍 I'd say go with one of those (I don't know whether it matters which one). I think it's good to keep `brew list` as fast as possible so while `Formula.installed.any?` is cleaner code-wise I'd prefer these.

Should the same thing be done with `Cask::Caskroom.path` for casks? Not sure which is faster. I didn't notice the same speed issue when running `Cask::Caskroom.casks` in `brew irb`, so it may not be a huge concern.